src/WorkbenchAccessManager.php

```diff
@@ -121,351 +153,251 @@ public function getActiveTree() {
-    $return = FALSE;
-    // Get the information from the account.
-    if (is_null($uid)) {
-      $uid = \Drupal::currentUser()->id();
-    }
-    $user = \Drupal::entityTypeManager()->getStorage('user')->load($uid);
-    if ($user->hasPermission('bypass workbench access')) {
-      $return = TRUE;
-    }
-    else {
-      // If the user is assigned to all the top-level sections, treat as admin.
-      $user_sections = $this->getUserSections($uid);
-      foreach (array_keys($this->getActiveTree()) as $root) {
-        $return = TRUE;
-        if (empty($user_sections[$root])) {
-          $return = FALSE;
+  public function checkTree($entity_sections, $user_sections) {
+    $tree = $this->getActiveTree();
+    $list = array_flip($user_sections);
+    foreach ($entity_sections as $section) {
+      // Simple check first: is there an exact match?
+      if (isset($list[$section])) {
+        return TRUE;
+      }
+      // Check for section on the tree.
+      foreach ($tree as $id => $info) {
+        if (isset($list[$section]) && isset($info[$section])) {
+          return TRUE;
+        }
+        // Recursive check for parents.
+        if (!empty($info[$section]['parents'])) {
+          $parents = array_flip($info[$section]['parents']);
+          // Check for parents.
+          foreach ($list as $uid => $data) {
+            if (isset($parents[$uid])) {
+              return TRUE;
+            }
+          }
         }
       }
     }
-    return $return;
+    return FALSE;
   }
 
   /**
-   * {@inheritdoc}
+   * @inheritdoc
    */
-  public function addUser($user_id, $sections = array()) {
-    $entity = \Drupal::entityManager()->getStorage('user')->load($user_id);
-    $values = $this->getUserSections($user_id, FALSE);
-    $new = array_merge($values, $sections);
-    $entity->set(WorkbenchAccessManagerInterface::FIELD_NAME, $new);
-    $entity->save();
+  public function flushFields() {
+    // Flush the field settings.
+    $node_types = $this->entityTypeManager->getStorage('node_type')->loadMultiple();
+    foreach ($node_types as $id => $type) {
+      $type->setThirdPartySetting('workbench_access', 'workbench_access_status', 0);
+      $type->save();
+      $fields['node'][$id] = '';
+    }
+    $this->configFactory->getEditable('workbench_access.settings')
+      ->set('fields', $fields)
+      ->save();
+    drupal_set_message($this->t('Field settings reset.'));
   }
 
   /**
    * {@inheritdoc}
+   *
+   * @deprecated To be removed before 8.x-1.0 - use
+   *   \Drupal\workbench_access\RoleSectionStorage::addRole() instead
    */
   public function addRole($role_id, $sections = array()) {
-    $settings = $this->loadRoleSections($role_id);
-    foreach ($sections as $id) {
-      $settings[$id] = 1;
-    }
-    $this->saveRoleSections($role_id, $settings);
+    trigger_error(__CLASS__ . '::' . __METHOD__ . ' is deprecated, please use \Drupal\workbench_access\RoleSectionStorage::' . __METHOD__ . ' instead', E_USER_DEPRECATED);
```

Comment:
Left these in and used the standard E_USER_DEPRECATED error so they're easy to find for anyone using the api.

Will write a change notice, we can keep this around for the next couple of alphas, then drop it.